tuf/api/metadata.py

```diff
@@ -0,0 +1,289 @@
+updates, and create and verify signatures.
+
+TODO:
+
+ * Add docstrings
+
+ * Finalize/Document Verify/Sign functions (I am not fully sure about expected
+   behavior)
+
+ * Validation (some thoughts ...)
+   - Avoid schema, see secure-systems-lab/securesystemslib#183
+   - Provide methods to validate JSON representation (at user boundary)
+   - Fail on bad json metadata in read_from_json method
+   - Be lenient on bad/invalid metadata objects in memory, they might be
+     work in progress. E.g. it might be convenient to create empty metadata
+     and assign attributes later on.
+   - Fail on bad json metadata in write_to_json method, but with option to
+     disable check as there might be a justified reason to write WIP
+     metadata to json.
+
+ * It might be nice to have short-cuts on the Metadata class to methods and
+   attributes of the contained Signed object. If we do this, we should only do
+   it on common methods/attributes (e.g. version, bump_version, expires,
+   bump_expiration)
+
+ * Similarly, it might be nice to have a generic Metadata.read_from_json that
+   can load any TUF role metadata and instantiate the appropriate object based
+   on the json '_type' field.
+
+
+"""
+# Imports.
+
+# 1st-party.
+from tuf.api.keys import KeyRing
+
+# 2nd-party.
+from abc import ABC, abstractmethod
+from datetime import datetime, timedelta
+from typing import Any, Dict, List, Optional
+
+import json
+import logging
+import tempfile
+
+# 3rd-party.
+
+from securesystemslib.formats import encode_canonical
+from securesystemslib.util import load_json_file, persist_temp_file
+from securesystemslib.storage import StorageBackendInterface
+from tuf.repository_lib import (
+    _get_written_metadata,
+    _strip_version_number,
+    generate_snapshot_metadata,
+    generate_targets_metadata,
+    generate_timestamp_metadata,
+)
+
+import iso8601
+import tuf.formats
+
+# Types.
+
+JsonDict = Dict[str, Any]
+
+# Classes.
+
+
+class Metadata(ABC):
+    def __init__(
+            self, signed: 'Signed' = None, signatures: list = None) -> None:
+        # TODO: How much init magic do we want?
+        self.signed = signed
+        self.signatures = signatures
+
+    def as_dict(self) -> JsonDict:
+        return {
+            'signatures': self.signatures,
+            'signed': self.signed.as_dict()
+        }
+
+    def __update_signature(self, signatures, keyid, signature):
+        updated = False
+        keyid_signature = {'keyid':keyid, 'sig':signature}
+        for idx, keyid_sig in enumerate(signatures):
+            if keyid_sig['keyid'] == keyid:
+                signatures[idx] = keyid_signature
+                updated = True
+        if not updated:
+            signatures.append(keyid_signature)
+
+    def sign(self, key_ring: KeyRing) -> JsonDict:
+        # FIXME: Needs documentation of expected behavior
+        signed_bytes = self.signed_bytes
+        signatures = self.__signatures
```

Comment:
Is there any intention behind` self.__signatures` and `self.signatures` both existing?

Reply:
I don't think so. I haven't touched the sign/verify functions yet, because I wasn't sure about the original intended behavior (that's why I left above comment there).

But I think we agreed that my plans for these functions as outlined in https://github.com/theupdateframework/tuf/pull/1060#issuecomment-660056376 seemed reasonable.